samples/api/swapchain_recreation/swapchain_recreation.cpp

```diff
@@ -0,0 +1,936 @@
+	PerFrame &frame      = submit_history[submit_history_index];
+	if (frame.submit_fence != VK_NULL_HANDLE)
+	{
+		vkWaitForFences(get_device_handle(), 1, &frame.submit_fence, true, UINT64_MAX);
+
+		// Reset/recycle resources, they are no longer in use.
+		recycle_fence(frame.submit_fence);
+		recycle_semaphore(frame.acquire_semaphore);
+		vkResetCommandPool(get_device_handle(), frame.primary_command_pool, 0);
+
+		// Destroy any garbage that's associated with this submission.
+		for (SwapchainObjects &garbage : frame.swapchain_garbage)
+		{
+			cleanup_swapchain_objects(garbage);
+		}
+		frame.swapchain_garbage.clear();
+
+		// Note that while the submission fence, the semaphore it waited on and the command
+		// pool its command was allocated from are guaranteed to have finished execution,
+		// there is no guarantee that the present semaphore is not in use.
+		//
+		// This is because the fence wait above ensures that the submission _before_ present
+		// is finished, but makes no guarantees as to the state of the present operation
+		// that follows.  The present semaphore is queued for garbage collection when
+		// possible after present, and is not kept as part of the submit history.
+		assert(frame.present_semaphore == VK_NULL_HANDLE);
+	}
+
+	frame.submit_fence      = get_fence();
+	frame.acquire_semaphore = get_semaphore();
+	frame.present_semaphore = get_semaphore();
+
+	if (frame.primary_command_pool == VK_NULL_HANDLE)
+	{
+		VkCommandPoolCreateInfo cmd_pool_info{VK_STRUCTURE_TYPE_COMMAND_POOL_CREATE_INFO};
+		cmd_pool_info.flags            = VK_COMMAND_POOL_CREATE_TRANSIENT_BIT;
+		cmd_pool_info.queueFamilyIndex = queue->get_family_index();
+		VK_CHECK(vkCreateCommandPool(get_device_handle(), &cmd_pool_info, nullptr, &frame.primary_command_pool));
+	}
+}
+
+void SwapchainRecreation::render(uint32_t index)
+{
+	PerFrame &frame = submit_history[submit_history_index];
+
+	VkCommandBuffer             cmd;
+	VkCommandBufferAllocateInfo cmd_buf_info{VK_STRUCTURE_TYPE_COMMAND_BUFFER_ALLOCATE_INFO};
+	cmd_buf_info.commandPool        = frame.primary_command_pool;
+	cmd_buf_info.level              = VK_COMMAND_BUFFER_LEVEL_PRIMARY;
+	cmd_buf_info.commandBufferCount = 1;
+	VK_CHECK(vkAllocateCommandBuffers(get_device_handle(), &cmd_buf_info, &cmd));
+
+	VkCommandBufferBeginInfo begin_info{VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO};
+	begin_info.flags = VK_COMMAND_BUFFER_USAGE_ONE_TIME_SUBMIT_BIT;
+	vkBeginCommandBuffer(cmd, &begin_info);
+
+	// Render the following with basic vkCmdClearAttachments calls:
+	// - A gray rectangle that scales with the size of the extent
+	// - A fixed size square with changing color based on FPS
+
+	VkClearValue black;
+	black.color = {{0, 0, 0, 1.0f}};
+
+	VkClearValue gray;
+	gray.color = {{0.5f, 0.5f, 0.5f, 1.0f}};
+
+	VkClearValue colorful;
+	colorful.color = {{
+	    (frame_number % 256) / 255.0f,
+	    ((frame_number + 63) % 256) / 255.0f,
+	    ((frame_number + 128) % 256) / 255.0f,
```

Comment:
above 3 lines narrow convert from uint32_t to float.  Static_cast should help.

Reply:
The conversion is from an integer in `[0, 256)` though, there's no actual narrowing happening.

Added `static_cast` nevertheless.